gateway-api/src/gateway_api/time_duration.rs

```diff
@@ -0,0 +1,315 @@
+use delegate::delegate;
+use regex::Regex;
+use std::time::Duration as stdDuration;
+use kube::core::Duration as k8sDuration;
+use std::fmt;
+use std::str::FromStr;
+
+/// gateway_api::Duration is a duration type where parsing and formatting obey
+/// GEP-2257. It uses kube::core::Duration for the heavy lifting of parsing
+/// but is based on std::time::Duration.
+///
+/// Per GEP-2257, when parsing a gateway_api::Duration from a string, the
+/// string must match ^([0-9]{1,5}(h|m|s|ms)){1,4}$ and is otherwise parsed
+/// the same way that Go's time.ParseDuration parses durations. When
+/// formatting a gateway_api::Duration as a string, zero-valued durations must
+/// always be formatted as "0s", and non-zero durations must be formatted to
+/// with only one instance of each applicable unit, greatest unit first.
+///
+/// The rules above imply that gateway_api::Duration cannot represent
+/// durations with sub-millisecond precision or times greater than
+/// 99999h59m59s999ms. Since there's no meaningful way in Rust to allow string
+/// formatting to fail, these conditions are checked instead when
+/// instantiating gateway_api::Duration.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct Duration(stdDuration);
+
+// Maximum duration that can be represented by GEP-2257, in milliseconds.
+const MAX_DURATION_MS: u128 = (((99999 * 3600) + (59 * 60) + 59) * 1_000) + 999;
+
+/// is_valid checks if a duration is valid according to GEP-2257. If it's not,
+///
+fn is_valid(duration: stdDuration) -> Result<(), String> {
+    // Check nanoseconds to see if we have sub-millisecond precision in
+    // this duration.
+    if duration.subsec_nanos() % 1_000_000 != 0 {
+        return Err("Cannot express sub-millisecond precision in GEP-2257".to_string());
+    }
+
+    // Check the duration to see if it's greater than GEP-2257's maximum.
+    if duration.as_millis() > MAX_DURATION_MS {
+        return Err("Duration exceeds GEP-2257 maximum 99999h59m59s999ms".to_string());
+    }
+
+    Ok(())
+}
+
+/// Converting from std::time::Duration to gateway_api::Duration is allowed,
+/// but we need to make sure that the incoming duration is valid according to
+/// GEP-2257.
+impl TryFrom<stdDuration> for Duration {
+    type Error = String;
+
+    fn try_from(d: stdDuration) -> Result<Self, Self::Error> {
+        // Check validity, and propagate any error if it's not.
+        is_valid(d)?;
+
+        // It's valid, so we can safely convert it to a gateway_api::Duration.
+        Ok(Duration(d))
+    }
+}
+
```

Comment:
For the sake of ergonomics and such, would it be useful to add a `impl TryFrom<k8sDuration> for Duration` here as well?

Reply:
Yeah, I think it likely would be.